module/irrecoverable/example_test.go

```diff
@@ -0,0 +1,136 @@
+package irrecoverable_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/lifecycle"
+)
+
+type TemporaryError struct {
+	message string
+}
+
+func (s TemporaryError) Error() string {
+	return s.message
+}
+
+func Example() {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	starts := 0
+	var componentFactory module.ComponentFactory
+	componentFactory = func() (module.Component, error) {
+		var err error
+		starts++
+		if starts > 1 {
+			err = errors.New("Fatal! No restarts")
+		} else {
+			err = TemporaryError{"Restart me!"}
+		}
+
+		return NewExampleComponent(starts, err), nil
+	}
+
+	var onError module.OnError
+	onError = func(err error, triggerRestart func()) {
+		// check to make sure it's safe to restart the component
+		var tmp TemporaryError
+		if errors.As(err, &tmp) {
+			fmt.Printf("Restarting component after fatal error: %v\n", err)
+			triggerRestart()
+			return
+		}
+
+		// it's not safe to restart, shutdown instead
+		fmt.Printf("An unrecoverable error occurred: %v\n", err)
+		// shutdown other components. it might make sense to just panic here
+		// depending on the circumstances
+		cancel()
+	}
```

Comment:
I'd love comments here highlighting:
- "this is the place to implement the behaviors you want, e.g. firing an alert to pagerduty, etc ..."
- "the shutdown of the component is handled for you (RunComponent)"
- "you may consider additional cleanup (e.g. badgerDB reset)"